api/src/main/java/io/opencensus/trace/TraceState.java

```diff
@@ -0,0 +1,173 @@
+    for (Entry entry : entries) {
+      if (entry.key.equals(key)) {
+        return entry.value;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns a {@link List} view of the mappings contained in this {@code TraceState}.
+   *
+   * @return a {@link List} view of the mappings contained in this {@code TraceState}.
+   */
+  public List<Entry> entryList() {
+    return entries;
+  }
+
+  /**
+   * Creates a TraceState by appending the extra entries to the {@code parent} in front of the
+   * key-value pairs list and removing duplicate entries.
+   *
+   * @param parent the parent {@code TraceState}.
+   * @param extraEntries the list of extra entries.
+   * @return a TraceState by appending the extra entries to the {@code parent}.
+   */
+  public static TraceState create(TraceState parent, List<Entry> extraEntries) {
+    Utils.checkNotNull(parent, "parent");
+    Utils.checkNotNull(extraEntries, "extraEntries");
+    if (extraEntries.size() == 0) {
+      return parent;
+    }
+
+    for (Entry entry : extraEntries) {
+      Utils.checkArgument(validateKey(entry.key), "Invalid key " + entry.key);
+      Utils.checkArgument(validateKey(entry.value), "Invalid value " + entry.value);
+    }
+
+    List<Entry> ret = new ArrayList<Entry>(extraEntries.size() + parent.entries.size());
+    ret.addAll(extraEntries);
+    for (Entry entry : parent.entries) {
+      boolean isPresent = false;
+      for (Entry entryExtra : extraEntries) {
+        if (entryExtra.key.equals(entry.key)) {
+          isPresent = true;
+          break;
+        }
+      }
+      if (!isPresent) {
+        ret.add(entry);
+      }
+    }
+
+    Utils.checkState(ret.size() <= MAX_KEY_VALUE_PAIRS, "Invalid size");
+
+    return new TraceState(Collections.unmodifiableList(ret));
+  }
+
+  @Immutable
+  public static final class Entry {
+    private final String key;
+    private final String value;
+
+    private Entry(String key, String value) {
+      this.key = key;
+      this.value = value;
+    }
+  }
+
+  // Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter, and
+  // can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+  // forward slashes /.
+  private static boolean validateKey(String key) {
+    if (key.length() > KEY_MAX_SIZE
+        || key.isEmpty()
+        || key.charAt(0) < 'a'
+        || key.charAt(0) > 'z') {
+      return false;
+    }
+    for (int i = 0; i < key.length(); i++) {
+      char c = key.charAt(i);
+      if (!(c >= 'a' && c <= 'z')
+          && !(c >= '0' && c <= '9')
+          && c != '_'
+          && c != '-'
+          && c != '*'
+          && c != '/') {
+        return false;
+      }
+    }
+    return false;
+  }
+
+  // Value is opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the range
+  // 0x20 to 0x7E) except comma , and =.
+  private static boolean validateValue(String value) {
```

Comment:
maybe add href links to the w3c spec thing that describes this (to make sure for example the spec isn't wider than this, which could create a problem reading data if not)